subprojects/core/src/main/java/org/gradle/initialization/DefaultProjectDescriptor.java

```diff
@@ -40,7 +40,8 @@
 
     public static final String BUILD_SCRIPT_BASENAME = "build";
 
-    private String name;
+    private final String implicitName; // project name derived from the containing folder
```

Comment:
Might be better to store a boolean instead of both names, as the implicit one is not used at all if an explicit name is set.